mixer/pkg/attribute/envoyProtoBag.go

```diff
@@ -0,0 +1,317 @@
+	reqMap["request.time"] = reformatTime(request.GetTime(), 0)
+	reqMap["request.useragent"] = http.GetHeaders()["user-agent"]
+	fillContextProtocol(reqMap, http.GetHeaders())
+
+	pb.reqMap = reqMap
+
+	scope.Debugf("Returning bag with attributes:\n%v", pb)
+
+	return pb
+}
+
+// AccessLogProtoBag returns an attribute bag from a StreamAccessLogsMessage
+// When you are done using the proto bag, call the Done method to recycle it.
+// num is the index of the entry from the message's batch to create a bag from
+func AccessLogProtoBag(msg *accesslog.StreamAccessLogsMessage, num int) *EnvoyProtoBag {
+	pb := envoyProtoBags.Get().(*EnvoyProtoBag)
+	reqMap := make(map[string]interface{})
+	var commonproperties alspb.AccessLogCommon
+
+	if httpLogs := msg.GetHttpLogs(); httpLogs != nil {
+		commonproperties = *httpLogs.GetLogEntry()[num].GetCommonProperties()
+		request := httpLogs.GetLogEntry()[num].GetRequest()
+		response := httpLogs.GetLogEntry()[num].GetResponse()
+
+		reqMap["request.time"] = reformatTime(commonproperties.GetStartTime(), 0)
+		reqMap["response.time"] = reformatTime(commonproperties.GetStartTime(), commonproperties.GetTimeToFirstUpstreamRxByte().Nanos)
+		reqMap["request.headers"] = attr.WrapStringMap(request.GetRequestHeaders())
+		reqMap["request.host"] = request.GetAuthority()
+		reqMap["request.method"] = request.GetRequestMethod().String()
+		reqMap["request.path"] = request.GetPath()
+		reqMap["request.scheme"] = request.GetScheme()
+		reqMap["request.useragent"] = request.GetUserAgent()
+		reqMap["response.headers"] = attr.WrapStringMap(response.GetResponseHeaders())
+		reqMap["response.code"] = int64(response.GetResponseCode().GetValue())
+		reqMap["response.size"] = int64(response.GetResponseBodyBytes())
+		reqMap["response.total_size"] = int64(response.GetResponseBodyBytes()) + int64(response.GetResponseHeadersBytes())
+		fillContextProtocol(reqMap, request.GetRequestHeaders())
+	} else if tcpLogs := msg.GetTcpLogs(); tcpLogs != nil {
+		commonproperties = *tcpLogs.GetLogEntry()[num].GetCommonProperties()
+		connection := tcpLogs.GetLogEntry()[num].GetConnectionProperties()
+
+		reqMap["context.time"] = reformatTime(commonproperties.GetStartTime(), 0)
+		reqMap["connection.received.bytes"] = int64(connection.GetReceivedBytes())
+		reqMap["connection.sent.bytes"] = int64(connection.GetSentBytes())
+		reqMap["context.protocol"] = "tcp"
+	}
+	reqMap["context.reporter.kind"] = "inbound"
+	fillAddress(reqMap, commonproperties.GetDownstreamLocalAddress(), "destination")
+	fillAddress(reqMap, commonproperties.GetDownstreamRemoteAddress(), "source")
+	tlsproperties := commonproperties.GetTlsProperties()
+	reqMap["destination.principal"] = tlsproperties.GetLocalCertificateProperties().GetSubjectAltName()[0].GetUri()
+	reqMap["source.principal"] = tlsproperties.GetPeerCertificateProperties().GetSubjectAltName()[0].GetUri()
+	reqMap["connection.requested_server_name"] = commonproperties.GetTlsProperties().GetTlsSniHostname()
```

Comment:
Here and throughout: TLS properties might be empty. Please add nil checks whenever you dereference nested properties.

Reply:
Done